src/fidasim.f90

```diff
@@ -1860,6 +1904,87 @@ elemental function lfs_divide(p1, real_scalar) result (p3)
 
 end function lfs_divide
 
+elemental function oo_add(p1, p2) result (p3)
+    !+ Defines how to add two [[CFPDTable]] types
+    type(CFPDTable), intent(in) :: p1,p2
+    type(CFPDTable)             :: p3
+
+    p3%nenergy   = p1%nenergy   + p2%nenergy
+    p3%nrays     = p1%nrays     + p2%nrays
+    p3%nsteps    = p1%nsteps    + p2%nsteps
+    p3%earray    = p1%earray    + p2%earray
+    p3%nactual   = p1%nactual   + p2%nactual
+    p3%daomega   = p1%daomega   + p2%daomega
+    p3%sightline = p1%sightline + p2%sightline
+
+end function oo_add
+
+elemental function oo_subtract(p1, p2) result (p3)
+    !+ Defines how to subtract two [[CFPDTable]] types
+    type(CFPDTable), intent(in) :: p1,p2
+    type(CFPDTable)             :: p3
+
+    p3%nenergy   = p1%nenergy   - p2%nenergy
+    p3%nrays     = p1%nrays     - p2%nrays
+    p3%nsteps    = p1%nsteps    - p2%nsteps
+    p3%earray    = p1%earray    - p2%earray
+    p3%nactual   = p1%nactual   - p2%nactual
+    p3%daomega   = p1%daomega   - p2%daomega
+    p3%sightline = p1%sightline - p2%sightline
+
+end function oo_subtract
+
+elemental function oos_multiply(p1, real_scalar) result (p3)
+    !+ Defines how to multiply two [[CFPDTable]] types
+    type(CFPDTable), intent(in) :: p1
+    real(Float64), intent(in)  :: real_scalar
+    type(CFPDTable)             :: p3
+
+    p3%nenergy   = p1%nenergy   * real_scalar
+    p3%nrays     = p1%nrays     * real_scalar
+    p3%nsteps    = p1%nsteps    * real_scalar
+    p3%earray    = p1%earray    * real_scalar
+    p3%nactual   = p1%nactual   * real_scalar
+    p3%daomega   = p1%daomega   * real_scalar
+    p3%sightline = p1%sightline * real_scalar
+
+end function oos_multiply
+
+elemental function soo_multiply(real_scalar, p1) result (p3)
+    !+ Defines how to multiply [[CFPDTable]] types by a scalar
+    type(CFPDTable), intent(in) :: p1
+    real(Float64), intent(in)  :: real_scalar
+    type(CFPDTable)             :: p3
+
+    p3 = p1*real_scalar
+
+end function soo_multiply
+
+elemental function oos_divide(p1, real_scalar) result (p3)
+    !+ Defines how to divide [[CFPDTable]] types by a scalar
+    type(CFPDTable), intent(in) :: p1
+    real(Float64), intent(in)  :: real_scalar
+    type(CFPDTable)             :: p3
+
+    p3 = p1*(1.d0/real_scalar)
+
+end function oos_divide
+
+pure subroutine oo_assign(p1, p2)
+    !+ Defines how to assign [[CFPDTable]] types to eachother
+    type(CFPDTable), intent(in)  :: p2
+    type(CFPDTable), intent(out) :: p1
+
+    p1%nenergy   = p2%nenergy
+    p1%nrays     = p2%nrays
+    p1%nsteps    = p2%nsteps
+    p1%earray    = p2%earray
+    p1%nactual   = p2%nactual
+    p1%daomega   = p2%daomega
+    p1%sightline = p2%sightline
+
+end subroutine oo_assign
+
```

Comment:
Are these definitions really necessary. I may of missed it but I don't think they are used anywhere. Also I question whether the operations make sense mathematically.

Reply:
Good question, I am not sure if this is necessary. If you look at the bottom of read_cfpd, dE is defined using earray. This is the only case where a member of ctable is used to define a different member of ctable.

Could you please explain what the operator block is doing? I guess I don't fully understand the cases in which it would be important.

Reply:
The operator functions just define the rules about how to do addition, subtraction, multiply, divide, and assignment for your defined types. The interface block is just registering your operator functions so that the compiler knows what to do when you use the infix operators (+/-/*/=/etc) with your types. In your case I think only the assign functions are needed but you will know for sure when you compile.

Reply:
That make better sense, thanks. I removed most of the functions and kept only the assign function.